src/searchForm/autoComplete/overlay/AutoCompleteOverlay.tsx

```diff
@@ -1,15 +1,27 @@
 import React from 'react'
-import AutoComplete, { AutoCompleteProps } from 'autoComplete'
 import Bullet, { BulletTypes } from 'bullet'
+import { AutoCompleteProps } from 'autoComplete'
 
-export const AutoCompleteOverlay = (props: AutoCompleteProps) => {
+export interface AutoCompleteOverlayProps {
+  readonly name: string
+  readonly autocompleteComponent: React.ReactElement<AutoCompleteProps>
+  readonly className?: Classcat.Class
+  readonly onSelect?: (obj: AutocompleteOnChange) => void
+}
+
+export const AutoCompleteOverlay = (props: AutoCompleteOverlayProps) => {
   const icon = (
     <div className="kirk-autoComplete-icon">
       <Bullet type={BulletTypes.SEARCH} />
     </div>
   )
 
-  return <AutoComplete {...props} inputAddon={icon} embeddedInSearchForm />
+  return React.cloneElement(props.autocompleteComponent, {
```

Comment:
I find cloning element to add props error prone because from the cloned component point of view it's hard to know where the props come from. You can't search for the component usages in the project to find where props are passed ([grep test](http://jamie-wong.com/2013/07/12/grep-test/)).

They must also declare those props optional in order to be used as `prop={<MyComponent />}` even if they are required for the component to work.

A solution would be to use the render prop pattern to delegate the rendering:

```tsx
export const AutoCompleteOverlay = (props: AutoCompleteOverlayProps) => {
  // Caller
  return props.renderAutocompleteComponent({
    ...props,
    autoFocus: true,
    inputAddon: icon,
    embeddedInSearchForm: true,
  })
}

// Usage
<AutoCompleteOverlay
  renderAutocompleteComponent={(props) => (
    // Props are applied explicitly
    <AutoCompleteExample {...props} placeholder="Leaving from" />
  )}

  // It is event more flexible as you apply the props as you want
  renderAutocompleteComponent={(props) => (
    // The span wrapper doesn't get the props
    <span>
      <AutoCompleteExample {...props}  placeholder="Leaving from" />
    </span>
  )}
/>
```

What do you think?

Reply:
Yes much nicer indeed. I'll do this in the next PR. Thanks!